profiling/energy-profiler/kokkos/kp_nvml_energy_consumption.cpp

```diff
@@ -0,0 +1,533 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos Energy Consumption Profiler
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+/**
+ * @file kp_nvml_energy_consumption.cpp
+ * @brief Kokkos Energy Consumption Profiler Tool using NVML.
+ *
+ * This tool measures energy consumption by tracking the cumulative energy
+ * values from NVML at the beginning and end of kernels, regions, and deep
+ * copies. It does not use a background daemon since the energy consumption is a
+ * cumulative counter that can be read directly when events occur.
+ */
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <chrono>
+#include <mutex>
+#include <iomanip>
+#include <cmath>
+#include <fstream>
+#include <unordered_map>
+#include <stack>
+#include <memory>
+
+#include "kp_core.hpp"
+#include "../provider/provider_nvml.hpp"
+#include "../common/filename_prefix.hpp"
+#include "../common/timer.hpp"
+#include "../tools/kernel_timer_tool.hpp"
+
+namespace KokkosTools {
+namespace EnergyConsumption {
```

Comment:
There seems to be a lot of repetition ... maybe we should focus on the one we ended up using at the end